### PR TITLE
feat(ObjectViewer): Improve design

### PR DESCRIPTION
**Pull request by talend-glorieux**
**What is the problem this PR is trying to solve?**
The ObjectViewer is really basic.

**What is the chosen solution to this problem?**
Improve the graphical rendering of the ObjectViewer.

**Please check if the PR fulfills these requirements**
- [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

### CODE REVIEW

File: output/components.eslint.txt [hunk 1]
```diff
@@ -15,16 +15,28 @@ The react/require-extension rule is deprecated. Please use the import/extensions
...
   4:2  error  Unexpected require()                                                                    global-require
 
+/home/travis/build/Talend/ui/packages/components/src/Icon/Icon.component.js
+  70:4  error  Visible, non-interactive elements should not have mouse or keyboard event listeners  jsx-a11y/no-static-element-interactions
+
 /home/travis/build/Talend/ui/packages/components/src/ObjectViewer/__snapshots__/config.js
   1:1  error  '@storybook/react' should be listed in the project's dependencies, not devDependencies  import/no-extraneous-dependencies
   4:2  error  Unexpected require()                                                                    global-require
 
+/home/travis/build/Talend/ui/packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js
```
Comment (acateland) [inline output/components.eslint.txt]: bunch of code style to fix

File: output/components.sasslint.txt [hunk 2]
```diff
@@ -3,11 +3,50 @@
 > sass-lint -v -q
 
 
+src/ObjectViewer/JSONLike/JSONLike.scss
```
Comment (acateland) [inline output/components.sasslint.txt]: more linting issues to fix

File: output/containers.eslint.txt [hunk 3]
```diff
@@ -32,7 +32,9 @@ The react/require-extension rule is deprecated. Please use the import/extensions
   2:1  error  'react-test-renderer' should be listed in the project's dependencies, not devDependencies  import/no-extraneous-dependencies
 
 /home/travis/build/Talend/ui/packages/containers/src/ObjectViewer/ObjectViewer.test.js
-  2:1  error  'enzyme' should be listed in the project's dependencies, not devDependencies  import/no-extraneous-dependencies
+    2:1  error  'enzyme' should be listed in the project's dependencies, not devDependencies  import/no-extraneous-dependencies
+  101:5  error  Expected indentation of 3 tabs but found 4                                    indent
```
Comment (acateland) [inline output/containers.eslint.txt]: don't forget those ;p

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 4]
```diff
@@ -24,14 +36,32 @@ export function NativeValue({ data, edit, onClick, onChange, jsonpath }) {
 	if (edit) {
 		return <input type={inputType} value={data} onChange={e => onChange(e, { jsonpath })} />;
 	}
+
+	const isSelectedLine = (selectedJsonpath && (selectedJsonpath === jsonpath));
```
Comment (acateland) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: question why not simply `const isSelectedLine = selectedJsonpath === jsonpath;`

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 5]
```diff
@@ -24,14 +36,32 @@ export function NativeValue({ data, edit, onClick, onChange, jsonpath }) {
...
+
+	function stopAndSelect(event) {
```
Comment (acateland) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: this function is created on each of its parent eval, maybe you can try to extract it.

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 6]
```diff
@@ -42,9 +72,10 @@ NativeValue.propTypes = {
 		PropTypes.string,
 	]),
 	edit: PropTypes.bool,
-	onClick: PropTypes.func,
+	onSelect: PropTypes.func,
```
Comment (acateland) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: looking at the implementation this function is required and so should be tagged as required here

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 7]
```diff
@@ -99,76 +165,151 @@ export function getDataInfo(data) {
...
+
+		if (tupleLabel && tupleLabel.length > 0) {
+			info.type = tupleLabel;
+		}
 	}
 
 	return info;
 }
 
+export function abstracter(acc, item) {
```
Comment (acateland) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: please add a bit of documentation to this function,

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 8]
```diff
@@ -81,9 +144,12 @@ LineItem.propTypes = {
 	onMouseOver: PropTypes.func,
 	children: PropTypes.node,
 	mouseOverData: PropTypes.object,  // eslint-disable-line react/forbid-prop-types
+	jsonpath: PropTypes.string,
+	selectedJsonpath: PropTypes.string,
+	onSelect: PropTypes.func,
 };
 
-export function getDataInfo(data) {
+export function getDataInfo(data, tupleLabel) {
```
Comment (acateland) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: please add a bit of documentation to this function,

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 9]
```diff
@@ -99,76 +165,151 @@ export function getDataInfo(data) {
...
+		return OBJECT_ABSTRACT;
+	}
+	if (acc.length > 0) {
+		return `${acc}, ${item}`;
+	}
+
+	return `${item}`;
+}
+
+export function getDataAbstract(data) {
```
Comment (acateland) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: please add a bit of documentation to this function,

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 10]
```diff
@@ -99,76 +165,151 @@ export function getDataInfo(data) {
...
-									opened={opened}
-									edited={edited}
-								/>
-							</li>
-						);
-					})}
-				</ul>
-			</span>
-		</LineItem>
+		<div>
```
Comment (acateland) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: i think this could be splitted a bit

File: packages/containers/src/ObjectViewer/ObjectViewer.container.js [hunk 11]
```diff
@@ -53,18 +61,29 @@ class ObjectViewer extends React.Component {
...
 		if (data.isOpened) {
 			newState = close(data.jsonpath, this.props.state);
 		} else if (data.isOpened === false) {
 			// we don't want to match on undefined as false
 			newState = open(data.jsonpath, this.props.state);
-		} else if (data.edit === false) {
+		}
+
+		if (newState) {
+			this.props.setState(newState);
```
Comment (acateland) [inline packages/containers/src/ObjectViewer/ObjectViewer.container.js]: since this use a previous state to compute the new one, use the functionnal way to setState
the whole upper block should fit in this function

https://facebook.github.io/react/docs/react-component.html#setstate
this.setState((prevState, props) => {
  return {counter: prevState.counter + props.step};
});

File: packages/containers/src/ObjectViewer/ObjectViewer.container.js [hunk 12]
```diff
@@ -53,18 +61,29 @@ class ObjectViewer extends React.Component {
...
+		}
+	}
+
+	onEdit(event, data) {
+		let newState;
```
Comment (acateland) [inline packages/containers/src/ObjectViewer/ObjectViewer.container.js]: same as above !

File: packages/containers/src/ObjectViewer/ObjectViewer.container.js [hunk 13]
```diff
@@ -81,6 +100,14 @@ class ObjectViewer extends React.Component {
 		this.props.setState(newState);
 	}
 
+	onSelect(event, data) {
+		const newState = select(
```
Comment (acateland) [inline packages/containers/src/ObjectViewer/ObjectViewer.container.js]: same as above

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 14]
```diff
@@ -99,76 +165,151 @@ export function getDataInfo(data) {
...
+
 	if (data === undefined) {
 		return null;
 	}
-	const info = getDataInfo(data);
+	const info = getDataInfo(data, props.tupleLabel);
 	const isNativeType = COMPLEX_TYPES.indexOf(info.type) === -1;
 	const isEdited = edited.indexOf(jsonpath) !== -1 && !!props.onChange;
-	const isOpened = opened.indexOf(jsonpath) !== -1 || props.onClick === noop;
+	const isOpened = opened.indexOf(jsonpath) !== -1;
```
Comment (ncomont) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: You can use `includes` instead

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 15]
```diff
@@ -204,28 +352,52 @@ Item.defaultProps = {
...
  */
 export function JSONLike({ onSubmit, ...props }) {
+	const rootIsObject = isObject(props.data);
+	const rootComputedLabel = rootIsObject ? props.rootLabel ? props.rootLabel : 'root' : null;
+
 	if (onSubmit) {
 		return (
 			<form
-				className={`tc-object-viewer ${theme.container}`}
+				className={`tc- object - viewer ${theme.container} `}
```
Comment (ncomont) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: Why did you add spaces in the class name ?

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 16]
```diff
@@ -204,28 +352,52 @@ Item.defaultProps = {
...
+				</TooltipTrigger>
 				<Item {...props} />
 			</form>
 		);
 	}
+
+	console.log('dbg props', props);
 	return (
-		<div className={`tc-object-viewer ${theme.container}`}>
+		<div className={`tc- object - viewer ${theme.container} `}>
```
Comment (ncomont) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: Same question

File: packages/components/src/ObjectViewer/ObjectViewer.component.js [hunk 17]
```diff
@@ -14,21 +14,24 @@ export const DISPLAY_MODES = {
...
-	case DISPLAY_MODES.FLAT:
-		return (<Table {...props} flat />);
-	case DISPLAY_MODES.TREE:
-		return (<JSONLike {...props} />);
-	case DISPLAY_MODES.LIST:
-		return (<List {...props} />);
-	default:
-		return (<JSONLike {...props} />);
+		case DISPLAY_MODES.TABLE:
+			return (<Table {...props} />);
```
Comment (ncomont) [inline packages/components/src/ObjectViewer/ObjectViewer.component.js]: Parenthesis are useless for single line JSX.

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 18]
```diff
@@ -99,76 +165,151 @@ export function getDataInfo(data) {
...
+
+						{info.keys.map((key, i) => {
+							const jpath = getJSONPath(key, jsonpath, info.type);
+							return (
+								<li key={i}>
+									<Item
+										{...props}
+										data={data[key]}
+										name={key}
+										jsonpath={jpath}
```
Comment (ncomont) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: Could be `jsonpath={getJSONPath(key, jsonpath, info.type)}` to avoid jpath allocation.

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 19]
```diff
@@ -204,28 +352,52 @@ Item.defaultProps = {
  * @param {object} props react
  */
 export function JSONLike({ onSubmit, ...props }) {
+	const rootIsObject = isObject(props.data);
+	const rootComputedLabel = rootIsObject ? props.rootLabel ? props.rootLabel : 'root' : null;
```
Comment (ncomont) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: This is a bit hard to read. Could you use a regular if/else ?

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js [hunk 20]
```diff
@@ -204,28 +352,52 @@ Item.defaultProps = {
...
+					<div className={theme.rootLabelOverflow}>
+						{rootComputedLabel}
+					</div>
+				</TooltipTrigger>
 				<Item {...props} />
 			</form>
 		);
 	}
+
+	console.log('dbg props', props);
```
Comment (ncomont) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: Forgotten `console.log` ?

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.scss [hunk 21]
```diff
@@ -2,20 +2,133 @@
...
+	}
+
 	li {
 		list-style-type: none;
+		font-size: 1rem;
+		display: flex;
+	}
+
+	button {
+		font-size : 1.4rem;
```
Comment (ncomont) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.scss]: Lint

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js (hunk 4 above)
Comment (sgendre) [reply packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: The point is to avoid having true for each row when those two are valued to undefined or null

File: packages/containers/src/ObjectViewer/ObjectViewer.container.js (hunk 11 above)
Comment (sgendre) [reply packages/containers/src/ObjectViewer/ObjectViewer.container.js]: prevState is not type consistent between Map and plain old js object which causes functionnal way of setState to fail while using defined operation function (select, open...) that are using map.set method
+ code consistency with other method operation functions

**Comment by sgendre** [comment]
@ncomont screenshot : The all opened state on default tree component story was a hack that make the component state inconsistent causing further problems while closing complex types; btw you can see that the default container story do not display an all opened treeview neither

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js (hunk 18 above)
Comment (sgendre) [reply packages/components/src/ObjectViewer/JSONLike/JSONLike.component.js]: Easier debug purpose

File: packages/components/src/Icon/Icon.component.js [hunk 22]
```diff
@@ -59,6 +59,27 @@ function Icon({ className, name, title, transform }) {
...
+			className,
+			SVG_TRANSFORMS[transform],
+		);
+		return (
+			// eslint doesn't recognizes the xlinkHref mention
+			// eslint-disable-next-line jsx-a11y/no-static-element-interactions
+			<a
+				xlinkHref="#"
+				onClick={onClick}
+				className={`tfd-svg-anchor ${theme.link}`}
```
[the remark this answers is not in the document]
Comment (frassinier) [reply packages/components/src/Icon/Icon.component.js]: Use classnames too

File: packages/components/src/Icon/Icon.component.js [hunk 23]
```diff
@@ -59,6 +59,27 @@ function Icon({ className, name, title, transform }) {
 		);
 		return (<i className={classes} {...accessibility} />);
 	}
+	if (onClick && name) {
+		const classname = classnames(
+			theme['svg-icon'],
```
Comment (frassinier) [inline packages/components/src/Icon/Icon.component.js]: please prefix css modules ones by tc

File: packages/components/src/Icon/Icon.component.js [hunk 24]
```diff
@@ -80,6 +101,7 @@ Icon.propTypes = {
 	name: PropTypes.string.isRequired,
 	title: PropTypes.string,
 	transform: PropTypes.oneOf(TRANSFORMS),
+	onClick: PropTypes.func,
```
Comment (frassinier) [inline packages/components/src/Icon/Icon.component.js]: what if onClick is missing?

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.scss [hunk 25]
```diff
@@ -2,20 +2,133 @@
 .container {
 	display: block;
 
+	svg {
+		padding: 2px;
```
Comment (frassinier) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.scss]: please, use rem and why this 2px?

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.scss [hunk 26]
```diff
@@ -2,20 +2,133 @@
...
+		margin-right: 5px;
```
Comment (frassinier) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.scss]: there are some variables to use from @talend/theme

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.scss [hunk 27]
```diff
@@ -2,20 +2,133 @@
...
 		text-transform: none;
 	}
 
 	.name:after {
-		content: ':';
+		content: ':  ';
+	}
+
+	sup {
+		top: -0.2em;
```
Comment (frassinier) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.scss]: Slash random number and slash random unit FTW

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.scss [hunk 28]
```diff
@@ -2,20 +2,133 @@
...
+		height: 9px;
+		width: 9px;
+		border-width: 1px;
+		background-color: rgba($wedgewood, 0.3);
+		color: $wedgewood;
+		font-family: 'Source Sans Pro';
```
Comment (frassinier) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.scss]: It's default one

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.scss [hunk 29]
```diff
@@ -2,20 +2,133 @@
...
-		content: ':';
+		content: ':  ';
+	}
+
+	sup {
+		top: -0.2em;
+		height: 9px;
+		width: 9px;
+		border-width: 1px;
+		background-color: rgba($wedgewood, 0.3);
```
Comment (frassinier) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.scss]: Define var on top of this file please

File: packages/components/src/ObjectViewer/JSONLike/JSONLike.scss [hunk 30]
```diff
@@ -2,20 +2,133 @@
...
+		font-family: 'Source Sans Pro';
+		font-size: 15px;
+		vertical-align: baseline;
+		color: $wedgewood;
+		display: inline;
+	}
+
+	.line-value-selected {
+		font-family: 'Source Sans Pro';
+		font-size: 15px;
```
Comment (frassinier) [inline packages/components/src/ObjectViewer/JSONLike/JSONLike.scss]: defined twice? something goes wrong

File: packages/theme/src/theme/_colors.scss [hunk 31]
```diff
@@ -109,3 +109,28 @@ $wild-sand: #F6F6F6;
 ///
 /// @type Color
 $shadow: rgba(0, 0, 0, 0.2);
+
+/// DataStreams app color.
```
Comment (frassinier) [inline packages/theme/src/theme/_colors.scss]: This is generic, remove this please.

Comment (frassinier) [reply packages/theme/src/theme/_colors.scss]: This should be defined by final apps

File: output/containers.eslint.txt [hunk 32]
```diff
@@ -33,7 +33,11 @@ The react/require-extension rule is deprecated. Please use the import/extensions
   12:90  error  'notifications' is missing in props validation                                             react/prop-types
 
 /home/travis/build/Talend/ui/packages/containers/src/ObjectViewer/ObjectViewer.test.js
-  2:1  error  'enzyme' should be listed in the project's dependencies, not devDependencies  import/no-extraneous-dependencies
+    2:1   error    'enzyme' should be listed in the project's dependencies, not devDependencies  import/no-extraneous-dependencies
+    3:10  error    'List' is defined but never used                                              no-unused-vars
+    3:16  error    'Map' is defined but never used                                               no-unused-vars
+   15:7   error    'selectedPath' is assigned a value but never used                             no-unused-vars
+  104:3   warning  Unexpected console statement                                                  no-console
```
Comment (ncomont) [inline output/containers.eslint.txt]: Please fix this

**Comment by ncomont** [comment]
There is a strange behaviour on IE11. The ObjectViewer become ligther when the user clicks on the background.

light

File: packages/components/src/Icon/Icon.component.js [hunk 33]
```diff
@@ -80,6 +101,7 @@ Icon.propTypes = {
 	name: PropTypes.string.isRequired,
 	title: PropTypes.string,
 	transform: PropTypes.oneOf(TRANSFORMS),
+	onClick: PropTypes.func.isRequired,
```
Comment (jmfrancois) [inline packages/components/src/Icon/Icon.component.js]: This is not required ...
